tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,404 @@
+class RiGLPruner(pruner.Pruner):
+  """
+  Implementation of the RiGL dynamic sparse training algorithm.
+  """
+
+  def __init__(self,
+      update_schedule=update_schedule.ConstantSchedule(0.1, 0),
+      sparsity=0.5,
+      block_size=(1,1),
+      block_pooling_type='AVG',
+      seed=0,
+      noise_std=0,
+      grow_init='zeros'
+    ):
+    """The logic for magnitude-based RiGL trained weight tensors as presented
+    in https://proceedings.icml.cc/static/paper_files/icml/2020/287-Paper.pdf.
+    At each iteration, according to the update_schedule, connections are grown
+    based on gradient information and dropped based on weight magnitude.
+
+    Args:
+      update_schedule: A `PruningSchedule` object that controls pruning rate
+        throughout training. The first arg is fraction of connections to drop
+        during the update, the second is the starting step to begin updating.
+      sparsity: layer specific sparsity at which the dynamic sparse training method uses
+      block_size: The dimensions (height, weight) for the block sparse pattern
+        in rank-2 weight tensors.
+      block_pooling_type: (optional) The function to use to pool weights in the
+        block. Must be 'AVG' or 'MAX'.
+      seed: assigned by PruningConfig (base added) for multiworker consistency in random processes.
+      grow_init: string ('zeros, 'random_normal', 'random_uniform', 'initial_value') way
+        to initialize new connections (i.e. for a connection that was dropped and then grown, 
+        this is to decide how these newly grown weights are initialized),
+    """
+    super(RiGLPruner, self).__init__()
+    self.target_sparsity = sparsity
+    self.update_schedule = update_schedule
+    self._block_size = block_size
+    self.block_pooling_type = block_pooling_type
+    self._seed = seed
+    self._noise_std = noise_std
+    valid_grow_inits = ('zeros', 'random_normal', 'random_uniform', 'initial_value')
+    try:
+      self._grow_init_method = grow_init.lower()
+    except:
+      raise ValueError(f"Check that the grow_init method is type str and one of {valid_grow_inits}")
+    if self._grow_init_method not in valid_grow_inits:
+      raise ValueError(f'The initialization for growing {grow_init} is not a valid option ({valid_grow_inits})')
+    # boolean for whether to reinitialize a connection that was dropped and regrown to 
+    # its original value, or to set it to the value specified by `grow_init` otherwise.
+    self._drop_regrow_reinit = True if self._grow_init_method == 'initial_value' else False 
+    
+  
+  def create_slots(self, optimizer, var):
+    base_dtype = var.dtype
+    _deterministic_initializer = sparse_utils.PermuteOnes(self.target_sparsity)
+    deterministic_initializer = functools.partial(_deterministic_initializer, dtype=base_dtype, seed=self._seed)
+    optimizer.add_slot(var, 'mask', initializer=deterministic_initializer)
+    optimizer.add_slot(var, 'initial_value', initializer=var.read_value())
```

Comment:
Add a todo: only do this if the configuration params require it. Otherwise we're increasing our memory usage for no good reason.

Also add a todo: see if this actually works in practice in a tf function as written. (it may or may not need to make a variable *then assign* rather than passing var.read_value() as the initializer.

Reply:
there is a test for this, it works in a tf function as written.